.changeset/moody-toys-dance.md

```diff
@@ -0,0 +1,31 @@
+---
+"@fluidframework/tree": minor
+"fluid-framework": minor
+---
+
+Allow constructing ArrayNodes from Maps and MapNodes from arrays when unambiguous.
+
+Since the types for ArrayNodes and MapNodes indicate they can be constructed from iterables,
+it should work, even if those iterables are themselves arrays or maps.
+To avoid this being a breaking change, a priority systems was introduced.
+ArrayNodes will only be implicitly constructable from JavaScript Map objects in contexts where no MapNodes are allowed.
+Similarly MapNodes will only be implicitly constructable from JavaScript Array objects in contexts where no ArrayNodes are allowed.
```

Comment:
Can we improve upon this even more in the future? You could have e.g. `Map<string, string> | Array<string>`, and we would be able to tell the difference between the two for an input like `[string]` (array) or `[[string, string]]` (map), so long as the array isn't empty. Or, is that too subtle of an edge case that it will confuse users, and it's better to simply keep the dividing line as it is? I'm inclined to think that it's too subtle, but I'm curious what you think.

Reply:
Yes. You can't tell as `[string, string][]` from a `number[]` in the case where they are empty, so in general we can't do that general improvement always, but we can do some cases of it (ex: all non-empty cases) in the future, and doing so will be a non-breaking change since it would convert cases which were ambiguous to working, and cases which had type errors on the children to type errors on the parent. Both those changes will be allowed.

Reply:
Note that the kind of change you mention (being more picky, causing error higher up the tree and less ambiguity) is generally safe, where the change I made here (allowing more options) is prone to causing new ambiguity issues which is what needed the extra care and priority thing

Reply:
Good observation, that makes sense. Thank you.